### ISSUE TITLE
Upgrade to ASP.NET Core 3.1

### ISSUE THREAD

**Issue by CrossBound** [PiranhaCMS/piranha.core#878]
I tried creating a brand new web app with asp.net core 3 and adding in Piranha, but it will not work. I believe it is the same issue as references here: https://github.com/aspnet/AspNetCore/issues/8254

During startup the app throws the exception below. Any idea when ASP.NET Core 3 will be supported?

```
Exception has occurred: CLR/System.TypeLoadException
An unhandled exception of type 'System.TypeLoadException' occurred in WebForm.dll: 'Could not load type 'Microsoft.AspNetCore.Mvc.MvcJsonOptions' from assembly 'Microsoft.AspNetCore.Mvc.Formatters.Json, Version=3.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'.'
   at ManagerModuleExtensions.AddPiranhaManagerOptions(IMvcBuilder builder)
   at WebForm.Startup.ConfigureServices(IServiceCollection services) in /media/data/CrossBound Technology/Clients/Portage/WebForm/WebForm/Startup.cs:line 24
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.InvokeCore(Object instance, IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.<>c__DisplayClass9_0.<Invoke>g__Startup|0(IServiceCollection serviceCollection)
   at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.Invoke(Object instance, IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.<>c__DisplayClass8_0.<Build>b__0(IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.GenericWebHostBuilder.UseStartup(Type startupType, HostBuilderContext context, IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.GenericWebHostBuilder.<>c__DisplayClass12_0.<UseStartup>b__0(HostBuilderContext context, IServiceCollection services)
   at Microsoft.Extensions.Hosting.HostBuilder.CreateServiceProvider()
   at Microsoft.Extensions.Hosting.HostBuilder.Build()
   at WebForm.Program.Main(String[] args) in /media/data/CrossBound Technology/Clients/Portage/WebForm/WebForm/Program.cs:line 10
```

**Comment by CrossBound** [comment]
P.S. I'm using the latest versions of Piranha from nuget.

**Comment by tidyui** [comment]
Hi there. There are several issues with running Piranha in .NET Core 3. First off, like you noticed, you have to replace the new JSON serializer with Newtonsoft as the new serializer can’t serialize/deserialize abstract classes. Secondly there’s a difference in how the change tracker of Entity Framework Core 3 works which breaks the application.

We have a separate branch where we’re almost done with the 3.0 upgrade but it’s decided when it will be released as it’s not backwards compatible.

Regards

**Comment by CrossBound** [comment]
Ok, thank you.

**Comment by jensbrak** [comment]
> Hi there. There are several issues with running Piranha in .NET Core 3. First off, like you noticed, you have to replace the new JSON serializer with Newtonsoft as the new serializer can’t serialize/deserialize abstract classes. Secondly there’s a difference in how the change tracker of Entity Framework Core 3 works which breaks the application.
> 
> We have a separate branch where we’re almost done with the 3.0 upgrade but it’s decided when it will be released as it’s not backwards compatible.
> 
> Regards
 The 
Excellent news, since I'd love to move up to 3 just for the sake of it. Looking forward to it, I

**Comment by distaula** [comment]
>  Secondly there’s a difference in how the change tracker of Entity Framework Core 3 works which breaks the application.
> 

Could you explain a little more about the Entity Framework issue?  I've been tinkering with a Blazor Frontend here: https://github.com/distaula/piranha.core/tree/initial-blazor and am curious what kind of issues I may run into.

**Comment by tidyui** [comment]
The difference is that in previous versions, adding an untracked item to a mapped collection of a tracked item would add the item into the change tracker, like this:

    var obj = db.Obj.FirstOrDefault(o => o.Id == id);
    obj.Items.Add(new Item());

    db.SaveChanges();

In `3.0` this seems to throw an exception as the state is wrong of the "item" unless it's manually added into it's `DbSet` as well.

I think the current errors mostly has to do with saving the application, so if you're building a Blazor `.NET Core 3.0` front-end that fetches data from a `.NET Core 2.2` backend with the Manager & API:s you could be fine :)

**Comment by rengert** [comment]
Because of bugfixes pushed to master, this would fix version 7.1. too ... i guess. When can we expect the Version 8 released? (the issue was, you mixed 2.2 und 3.0 assemplies during nuget package generation with not expected side effects). Just let me know.

**Comment by tidyui** [comment]
Hi @rengert 8.0 will be released in the next day or two. We’re just finalizing some things
